in-progress/8403-blobbity-boo.md

```diff
@@ -0,0 +1,348 @@
+
+Thankfully, there is a more efficient way, already implemented by @iAmMichaelConnor in the [`blob-lib`](https://github.com/iAmMichaelConnor/blob-lib) repo and [`blob`](https://github.com/AztecProtocol/aztec-packages/tree/master/noir-projects/noir-protocol-circuits/crates/blob) crate in aztec-packages.
+
+Our goal is to efficiently show that our tx effects accumulated in the rollup circuits are the same $d_i$ s in the blob committed to by $C$ on L1. To do this, we can provide an *opening proof* for $C$. In the circuit, we evaluate the polynomial at a challenge value $z$ and return the result: $p(z) = y$. We then construct a [KZG proof](https://dankradfeist.de/ethereum/2020/06/16/kate-polynomial-commitments.html#kate-proofs) in typescript of this opening (which is actually a commitment to the the quotient polynomial $q(x)$), and verify it on L1 using the [point evaluation precompile](https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile) added as part of EIP-4844. It has inputs:
+
+- `versioned_hash`: The `blobhash` for this $C$
+- `z`: The challenge value
+- `y`: The claimed evaluation value at `z`
+- `commitment`: The commitment $C$
+- `proof`: The KZG proof of opening
+
+It checks:
+
+- `assert kzg_to_versioned_hash(commitment) == versioned_hash`
+- `assert verify_kzg_proof(commitment, z, y, proof)`
+
+As long as we use our tx effect fields as the $d_i$ values inside the circuit, and use the same $y$ and $z$ in the public inputs of the Honk L1 verification as input to the precompile, we have shown that $C$ indeed commits to our data. Note: I'm glossing over some details here which are explained in the links above (particularly the 'KZG Proof' and 'host of benefits' links).
+
+But isn't evaluating $p(z)$ in the circuit also a bunch of very slow wrong-field arithmetic? No! Well, yes, but not as much as you'd think!
+
+To evaluate $p$ in evalulation form at some value not in its domain (i.e. not one of the $\omega^i$ s), we use the [barycentric formula](https://dankradfeist.de/ethereum/2021/06/18/pcs-multiproofs.html#evaluating-a-polynomial-in-evaluation-form-on-a-point-outside-the-domain):
+
+$$p(z) = A(z)\sum_{i=0}^{4095} \frac{d_i}{A'(\omega^i)} \frac{1}{z - \omega^i}.$$
+
+What's $A(x)$, you ask? Doesn't matter! One of the nice properties we get by defining $p$ as an interpolation over the roots of unity, is that the above formula is simplified to:
+
+$$p(z) = \frac{z^{4096} - 1}{4096} \sum_{i=0}^{4095} \frac{d_i\omega^i}{z - \omega^i}.$$
+
+We can precompute all the $\omega^i$, $-\omega^i$ s and $4096^{-1}$, the $d_i$ s are our tx effects, and $z$ is the challenge point (discussed more below). This means computing $p(z)$ is threoretically 4096 wrong-field multiplications and 4096 wrong-field divisions, far fewer than would be required for BLS12-381 elliptic curve operations.
```

Comment:
Under the current design, we wouldn't actually be computing `p(z)` in a circuit right?

But could/should we?

I.e. is it possible to effectively perform `verify_kzg_proof(commitment, z, y, proof)` in a circuit, and then have the user provide `commitment`, `y`, and `proof`. 

This convinces the circuit that any blob with that commitment has all the tx effects.

Then on L1 we:
1. store the commitments of the blobs at the time the block is proposed
2. match the stored commitments against the user provided ones when circuit verified the opening

That convinces L1 that the blobs the circuit proved against matched the ones that were previously published.

Reply:
@MirandaWood mentioned this as a possible approach to me recently. But she has dismissed it (and I agree) that computing a BLS12-381 pairing inside a Noir circuit would be many many constraints and very difficult to audit.

Reply:
> Under the current design, we wouldn't actually be computing p(z) in a circuit right

The barycentric formula is a way of computing p(z), so technically we _are_ computing it in the circuit with this proposal